storage/rocksdb/clone/common.h

```diff
@@ -0,0 +1,751 @@
+
+#ifndef CLONE_COMMON_H_
+#define CLONE_COMMON_H_
+
+#include <sys/types.h>
+#include <atomic>
+#include <cstddef>
+#include <cstdint>
+#include <ctime>
+#include <functional>
+#include <string>
+#include <unordered_set>
+
+#include "my_dbug.h"
+#include "my_inttypes.h"
+#include "mysql/plugin.h"
+#include "mysql/psi/mysql_cond.h"
+#include "mysql/psi/mysql_mutex.h"
+#include "mysql/psi/mysql_rwlock.h"
+#include "mysqld_error.h"
+
+#include "../rdb_buff.h"
+#include "../rdb_psi.h"
+
+namespace myrocks {
+
+namespace clone {
+
+constexpr char checkpoint_name_prefix[] = ".clone_checkpoint-";
+constexpr char in_place_temp_datadir[] = ".rocksdb.cloned";
+constexpr char in_place_temp_wal_dir[] = ".rocksdb.wal.cloned";
+constexpr char in_progress_marker_file[] = ".clone_in_progress";
+
+constexpr auto dir_mode = S_IRWXU;
+
+constexpr std::uint64_t invalid_file_id =
+    std::numeric_limits<std::uint64_t>::max();
+constexpr my_off_t invalid_file_offset = std::numeric_limits<my_off_t>::max();
+constexpr char invalid_file_name[] = "";
+
+enum class metadata_type : std::uint32_t {
+  LOCATOR_V1,
+  FILE_NAME_MAP_V1,
+  FILE_CHUNK_V1,
+  ADD_ESTIMATE_V1,
+  MAX_VALID_TYPE = ADD_ESTIMATE_V1,
+  INVALID
+};
+
+// The header starts every metadata packet. Format:
+// uint32_t     version         currently constant 1; will be bumped if the
+//                              header format will change
+// uint32_t     payload_len     the payload length without this header
+// uint32_t     type            one of the valid metadata_type enumerators
+// This format happens to be identical to InnoDB clone metadata header format at
+// the moment but there is no requirement for it to be this way.
+//
+// This metadata packet is not associated with any data packet.
+class [[nodiscard]] metadata_header final {
+ public:
+  constexpr metadata_header(std::uint32_t payload_len_, metadata_type type_)
+      : payload_len{payload_len_}, type{type_} {}
+
+  void serialize(myrocks::Rdb_string_writer &buf) const {
+    assert(!is_invalid());
+    assert(buf.is_empty());
+
+    buf.write_uint32(version);
+    buf.write_uint32(payload_len);
+    buf.write_uint32(static_cast<std::underlying_type_t<decltype(type)>>(type));
+
+    assert(buf.get_current_pos() == length);
+  }
+
+  [[nodiscard]] static auto deserialize(Rdb_string_reader &buf) {
+    std::uint32_t read_version;
+    if (buf.read_uint32(&read_version)) return invalid();
+    if (read_version != version) return invalid();
+    std::uint32_t read_payload_len;
+    if (buf.read_uint32(&read_payload_len)) return invalid();
+    std::uint32_t read_metadata_type;
+    if (buf.read_uint32(&read_metadata_type)) return invalid();
+    if (read_metadata_type >
+        static_cast<std::uint32_t>(metadata_type::MAX_VALID_TYPE))
+      return invalid();
+
+    metadata_header result{read_payload_len,
+                           static_cast<metadata_type>(read_metadata_type)};
+    assert(!result.is_invalid());
```

Comment:
I realize is_invalid() is used everywhere, but it might be easier to follow the code with is_valid(). Following !is_invalid() logic requires negating a negative.

Reply:
Good idea

Reply:
Fixed, squashed with other changes, rebased & pushed. With Valgrind fixes merged the last prerequisite commit still pending is https://github.com/facebook/mysql-5.6/pull/1254